README.md

```diff
@@ -1,32 +1,122 @@
+Core controller framework used by [Wrangler](https://github.com/rancher/wrangler) and [Norman](https://github.com/rancher/norman).
 
-#### Package `client`
+# Structure of this Document
+This document was written with two audiences in mind: someone who is new to Kubernetes controllers and someone who is not new to Kubernetes controllers but is new to lasso.
 
-Generic form of the generated clientsets.
+If you fall into the category of "Someone who is not new to Kubernetes controllers" you may skip to [Lasso's Value Proposition](#lassos-value-proposition).
 
-#### Package `controller`
+If you are new to Kubernetes controllers it is recommended you do not skip the any sections.
 
-The `Controller` is a simple approach that wraps the standard pattern of worker goroutines and workqueue.  The package
-depends on the `cache` and `client` removing the need for generated clientsets, informers, and listers.
+# Kubernetes Controllers
+_Note:_ This will be a very brief crash course. We encourage you to look up other resources regarding the topic, there are many!
 
-The `SharedController` type is an extension of `Controller` that adds the ability to register multiple handlers on a
-shared controller instance.
+From [the Kubernetes Documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)
+> In Kubernetes, a controller is a control loop that watches the shared state of the cluster through the apiserver and makes changes attempting to move the current state towards the desired state.
 
-#### Simple Example
+The word "controller" get used more loosely than what the above definition might have you believe. In our own words: Controllers are loop handlers that act in response to Kubernetes resource events. We also use the word "controller" to refer to the entire structure that is responsible for registering this loop handler and providing it the resources it needs. In short, controllers are structures that bundle everything necessary to interact with and operate on Kubernetes types.
+
+In order to achieve the above goal, a few sub-goals need to be met:
+* the controller needs to know about resource type events
+* the controller needs to be passed relevant details regarding the event to a handler
+* (optional) the controller __may__ want to interact with Kubernetes resource types
+
+There are many frameworks to help achieve the above goals. Lasso is one of them. What most of these frameworks have in common, including lasso, is that they use functionality from client-go to achieve these goals. You could achieve these goals by using just client-go. For the remainder of this section, we will fly through how a controller is put together just using client-go concepts.
+
+An informer is used to achieve the above goals needed for a Kubernetes Controller. An informer possesses an event handler and a set of indexers. The underlying structs are different but a pseudo struct is shown below to illustrate this:
```

Comment:
Before introducing how an informer is composed, it would help highlighting its role/function/goal.

Actually, since a lot of concepts are introduced, it would help adding definitions for each new jargon term, with examples eg.:

A **resource** is anything Kubernetes manages - such as pods, ingresses, jobs and so on.

An **event** is a change to a resource - such as an addition, an update or a deletion.

An **informer** is a Go object that observes a set of resources through the Kubernetes API and reacts to events pertaining to them. It also keeps a local cache of its objects, to avoid to load the API excessively. You could have your own informer to react to changes in basic Kubernetes resources, or your own.

An **indexer** is a component inside of informers. Its job is to maintain copies of resources temporarily - for example as caches or work queues. As the name implies, indexers keep indices on their resources for fast access.

...and so on

Reply:
I feel like indexers are further explained here https://github.com/rancher/lasso/pull/38/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R63 and informers are better explained here https://github.com/rancher/lasso/pull/38/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R65. The linked image should also help clarify here https://github.com/rancher/lasso/pull/38/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R69.

Reply:
Otherwise, I want to keep this doc more on the lean side. I want to give just enough info if you haven't been exposed to controllers but I don't think this is the place for a general k8s 101.

Reply:
Added a note here: https://github.com/rancher/lasso/pull/38/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R38.

Reply:
OK, so, I did not write that comment to suggest broadening the scope of the content, but rather to give an example on how adding a bit more structure to the prose could help making it easier to scan for keywords.

A beginner would read the paragraph and extract all the important information. Then go forward with the reading (and perhaps the coding), inevitably forget part of that and at some later point ask "what was an indexer again?" - having a succint, isolated paragraph with **indexer** in bold somewhere helps more than plain prose in answering that question.

That said I completely trust you in knowing readers better and picking the most appropriate final form 👍